File: java/236.lowest-common-ancestor-of-a-binary-tree.java
/*
 * @lc app=leetcode id=236 lang=java
 *
 * [236] Lowest Common Ancestor of a Binary Tree
 */
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode(int x) { val = x; }
 * }
 */
class Solution {

    public TreeNode lowestCommonAncestor(TreeNode root, TreeNode p, TreeNode q) {
        if(root == null || root == p || root == q)  return root;
        TreeNode left = lowestCommonAncestor(root.left, p, q);
        TreeNode right = lowestCommonAncestor(root.right, p, q);
        if(left != null && right != null)   return root;
        return left != null ? left : right;
    }

    /*
    public TreeNode lowestCommonAncestor(TreeNode root, TreeNode p, TreeNode q) {
        if (p == q) return p;
        Stack<TreeNode> sp = new Stack<>();
        Stack<TreeNode> sq = new Stack<>();
        if (!findNode(root, p, sp) || !findNode(root, q, sq)) return null;
        while (sp.size() < sq.size()) sq.pop();
        while (sq.size() < sp.size()) sp.pop();
        TreeNode res = sp.pop();
        while (res != sq.pop()) res = sp.pop();
        return res;
    }

    private boolean findNode(TreeNode root, TreeNode target, Stack<TreeNode> path) {
        if (root == null) return false;
        path.push(root);
        if (root == target) return true;
        if (!findNode(root.left, target, path) && !findNode(root.right, target, path)) {
            path.pop();
            return false;
        }
        return true;
    }
    */

}

File: go/236.lowest-common-ancestor-of-a-binary-tree.go
/*
 * @lc app=leetcode id=236 lang=golang
 *
 * [236] Lowest Common Ancestor of a Binary Tree
 */

// @lc code=start
/**
 * Definition for TreeNode.
 * type TreeNode struct {
 *     Val int
 *     Left *ListNode
 *     Right *ListNode
 * }
 */
func lowestCommonAncestor(root, p, q *TreeNode) *TreeNode {
	pp, pq := []*TreeNode{}, []*TreeNode{}
	dfs(root, p, &pp)
	dfs(root, q, &pq)
	if len(pp) > len(pq) {
		pp, pq = pq, pp
	}
	for i := 1; i < len(pp); i++ {
		if pp[i] != pq[i] {
			return pp[i-1]
		}
	}
	return pp[len(pp)-1]
}

func dfs(root, node *TreeNode, path *[]*TreeNode) bool {
	if root == nil {
		return false
	}
	*path = append(*path, root)
	if root == node {
		return true
	}
	exist := dfs(root.Left, node, path) || dfs(root.Right, node, path)
	if exist {
		return true
	}
	*path = (*path)[:len(*path)-1]
	return false
}
// @lc code=end

